Refactor the driver to support colocated option

ashao opened the pull request:
The driver for running the NCAR EKE example has been rewritten to
be a bit cleaner and more inline with the other examples in the
SmartSim Zoo. Additionally, a runtime option has been included
to attach a colocated database to the model instead of using
a distributed cluster.

ashao committed:
Refactor the driver to support colocated option

The driver for running the NCAR EKE example has been rewritten to
be a bit cleaner and more inline with the other examples in the
SmartSim Zoo. Additionally, a runtime option has been included
to attach a colocated database to the model instead of using
a distributed cluster.

ashao committed:
MOM6 driver now works with colocated

ashao committed:
Reorganize repo, update submodule, and add additional modules

ashao committed:
Merge branch 'master' of github.com:CrayLabs/NCAR_ML_EKE into add_colocated_db

ashao committed:
Missed a commit

ashao commented:
Driver is ready for review.

The MOM6 tag was not made for the original one submission since it points to your fork. However, it should still be available if someone were to recursively clone the jcs_iniitial tag.

I'll move onto the instructions and notebook next week.

Spartee commented on MOM6_config/OM4_025/.ipynb_checkpoints/MOM_override-checkpoint:
```diff
@@ -0,0 +1,9 @@
+! Blank file in which we can put "overrides" for parameters
+!VERBOSITY = 9
+EKE_SOURCE='silc'
+SILC_MODEL='ncar_ml_eke.gpu.pt'
```
change all these refereces to ``SILC``

ashao replied:
This should be deleted...it's a super old notebook thing.

Spartee commented on driver/driver.py:
```diff
@@ -20,81 +9,266 @@
+        batch_args = mom6_batch_args
+    )
+
+    mom6_run_args = {
```
use ``set_tasks`` and ``set_tasks_per_node``

Spartee commented on driver/driver.py:
```diff
@@ -20,81 +9,266 @@
+    mom6_run_args = {
+        "nodes"  : args.nodes_per_member,
+        "ntasks" : args.nodes_per_member*args.ppn_per_member,
+        "exclusive": None
```
when the ``RunSettings.set`` method is put in, use that here and use launcher==slurm as the condition.

Spartee commented on driver/driver.py:
```diff
@@ -20,81 +9,266 @@
+        )
+
+def create_distributed_orchestrator( args ):
+    orchestrator = Orchestrator(
```
``exp.create_database``

Spartee commented on driver/driver.py:
```diff
@@ -20,81 +9,266 @@
+def driver( ensemble_args, orchestrator_args, mom6_args):
+    log_to_file("./driver.log", log_level='developer')
+
+    experiment = Experiment("AI-EKE-MOM6", launcher="slurm")
```
``launcher=auto``

Spartee commented on driver/driver.py:
```diff
@@ -20,81 +9,266 @@
+    experiment.generate( *experiment_entities, overwrite=True )
+    experiment.start( *experiment_entities, summary=True )
+    print(experiment.summary())
+    experiment.stop()
```
this does nothing.

Spartee commented on driver/driver.py:
```diff
@@ -20,81 +9,266 @@
+    """
+
+    mom6_batch_args = {
+        "C":args.ensemble_node_features,
```
this is just about the only part of the script that will not work across all launchers. I would suggest changing the CLI to allow passing in ``--batch_arg=`arg=value``` or just define a structure at the top that contains these. that way it can at least be added to without removing this code.

Spartee commented on driver/driver.py:
```diff
@@ -20,81 +9,266 @@
+    )
+
+    mom_ensemble.attach_generator_files(
+        to_configure=glob("../MOM6_config/configurable_files/*"),
```
we should be able to support directories here now, right @MattToast ?

Spartee commented on driver/driver.py:
```diff
@@ -20,81 +9,266 @@
+        model.colocate_db(
+            args.orchestrator_port,
+            ifname=args.orchestrator_interface,
+            limit_app_cpus=False
```
this should be a configurable parameter.

Spartee replied:
I see you aren't using db_cpus which means the database is only getting 1 thread/core each run. I would recommend adding that in and making it a configurable parameter.

Spartee commented on driver/driver.py:
```diff
@@ -20,81 +9,266 @@
+        interface = args.orchestrator_interface,
+        db_nodes = args.orchestrator_nodes,
+        time="1:00:00",
+        threads_per_queue=4,
```
this is too high for a default. go with ``2``. remeber this is *per queue*

Spartee commented on driver/driver.py:
```diff
@@ -20,81 +9,266 @@
+        db_nodes = args.orchestrator_nodes,
+        time="1:00:00",
+        threads_per_queue=4,
+        launcher='slurm',
```
no need to pin launcher, use ``create_database`` and let the auto detection do its job.

Spartee commented on driver/driver.py:
```diff
@@ -20,81 +9,266 @@
+
+if __name__ == "__main__":
+    import argparse
+    parser = argparse.ArgumentParser(
```
this is ALOT of arguments. consider using something like fire (which we use in the scaling tests). it will automatically create these descriptions and what not based on docstrings. 

It will also allow you to pass in dictionaries which will be useful in making this work accross systems.

Spartee reviewed (changes requested):
Overall, great getting this updated. My biggest comment is with a little work, this will work accross all the launchers. the comment about using Fire may make that easier too. let me know if you have questions.

ashao committed:
Refactor driver for Fire CLI and cleanup INPUT

The driver for both colocated and clustered databases have been
refactored to be a bit more cleaner by using the capabilities of
Fire CLI. This also will make it easier for online analysis to be
done since we can return the configured experiment and allow it
to be run from a jupyter notebook.

ashao committed:
Add notebook to demonstrate online analysis and visualization

ashao committed:
Update notebook

ashao committed:
Remove old .ipynb checkpoint directory

Spartee commented on driver/driver.py:
```diff
@@ -20,81 +9,318 @@
-print(experiment.summary())
+if __name__ == "__main__":
+    import fire
+    log_to_file("./driver.log", log_level='developer')
```
name this something about MOM6. info for log_level

Spartee reviewed (approved):
Approved pending update to OA notebook. GOOD STUFF

ashao committed:
Fix up the onlne analysis notebook and remove the old SmartSim one

ashao committed:
Last small changes from review
